server/src/main/java/org/elasticsearch/common/xcontent/ChunkedToXContentBuilder.java

```diff
@@ -0,0 +1,168 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.common.xcontent;
+
+import org.elasticsearch.common.CheckedBiConsumer;
+import org.elasticsearch.core.Nullable;
+import org.elasticsearch.xcontent.ToXContent;
+import org.elasticsearch.xcontent.XContentBuilder;
+
+import java.io.IOException;
+import java.util.Iterator;
+import java.util.function.BiConsumer;
+import java.util.function.Consumer;
+import java.util.stream.Stream;
+
+/**
+ * A fluent builder to create {@code Iterator&lt;ToXContent&gt;} objects
+ */
+public class ChunkedToXContentBuilder implements Iterator<ToXContent> {
+
+    public static ChunkedToXContentBuilder builder(ToXContent.Params params) {
+        return new ChunkedToXContentBuilder(params);
+    }
+
+    private final ToXContent.Params params;
+    private final Stream.Builder<ToXContent> builder = Stream.builder();
+    private Iterator<ToXContent> iterator;
+
+    public ChunkedToXContentBuilder(ToXContent.Params params) {
+        this.params = params;
+    }
+
+    public ChunkedToXContentBuilder startObject() {
+        builder.add((b, p) -> b.startObject());
+        return this;
+    }
+
+    public ChunkedToXContentBuilder startObject(String name) {
+        builder.add((b, p) -> b.startObject(name));
+        return this;
+    }
+
+    public ChunkedToXContentBuilder endObject() {
+        builder.add((b, p) -> b.endObject());
+        return this;
+    }
+
+    public ChunkedToXContentBuilder startArray() {
+        builder.add((b, p) -> b.startArray());
+        return this;
+    }
+
+    public ChunkedToXContentBuilder startArray(String name) {
+        builder.add((b, p) -> b.startArray(name));
+        return this;
+    }
+
+    public ChunkedToXContentBuilder endArray() {
+        builder.add((b, p) -> b.endArray());
+        return this;
+    }
+
+    public ChunkedToXContentBuilder array(String name, String... values) {
+        builder.add((b, p) -> b.array(name, values));
+        return this;
+    }
+
+    public ChunkedToXContentBuilder field(String name, long value) {
```

Comment:
Rather than duplicating every method on `XContentBuilder`, we could have a single method
```
ChunkedToXContentBuilder append(Consumer<XContentBuilder> consumer)
```

That allowed a generic set of operations:
```
.append(b -> b.startObject("obj"))
.append(b -> b.field("foo", 100L))
.append(b -> b.field("bar", "val"))
.append(XContentBuilder::endObject)
```
but for ease-of-use, it's probably best to duplicate all the methods  here